src/lib/expandApplyAtRules.js

```diff
@@ -35,6 +37,94 @@ function prefix(context, selector) {
   return typeof prefix === 'function' ? prefix(selector) : prefix + selector
 }
 
+function* pathToRoot(node) {
+  yield node
+  while (node.parent) {
+    yield node.parent
+    node = node.parent
+  }
+}
+
+function shallowClone(node, overrides = {}) {
+  let children = node.nodes
+  node.nodes = []
+
+  const tmp = node.clone(overrides)
+
+  node.nodes = children
+
+  return tmp
+}
+
+/**
+ * @param {import('postcss').Node} node
+ */
+function structuralCloneOfNode(node) {
+  for (let parent of pathToRoot(node)) {
+    if (node === parent) {
+      continue
+    }
+
+    if (parent.type === 'root') {
+      break
+    }
+
+    node = shallowClone(parent, {
+      nodes: [node],
+    })
+  }
+
+  return node
+}
+
+/**
+ * @param {import('postcss').Root} root
+ */
+function buildLocalApplyCache(root, context) {
+  /** @type {ApplyCache} */
+  let cache = new Map()
+
+  let reservedBits = 0n
+  let tmp = context.layerOrder.utilities >> 3n
```

Comment:
Question, what's the reason for using `context.layerOrder.utilities >> 3n` instead of `context.layerOrder.user` instead?

Because I think the `>> 3n` was used because of this code:
```js
  context.layerOrder = {
    defaults: (1n << reservedBits) << 0n,
    base: (1n << reservedBits) << 1n,
    components: (1n << reservedBits) << 2n,
    utilities: (1n << reservedBits) << 3n, // The inverse of this part
    user: (1n << reservedBits) << 4n,
  }
```

and I wonder if it is save / better to just use the `context.layerOrder.user` since you want to put everything in that bucket.

Reply:
It's true that this number will be larger but I think that is fine. Eventually we still want to do the heuristic based sorting and then we just require the order of each layer and within that layer we can sort based on heuristic which should eventually clean up a lot of the BigInt code.

Reply:
I don't remember why I was using utilities instead of `user`. I think it was because there was an ordering issue in the first iteration but that's no longer a problem now. Updated.